gh-129515: Clarify syntax error messages for conditional expressions

sergey-miryanov opened the pull request:
<!-- gh-issue-number: gh-129515 -->
* Issue: gh-129515
<!-- /gh-issue-number -->

Along the suggestion from the original issue, I have added the following rules and messages:
The following expressions will be handled by `a=disjunction 'if' b=disjunction 'else' c=invalid_ifexp_orelse_stmt` and generate the message "Syntax Error: statement given where 'orelse' expression required" instead of just "Syntax Error: invalid syntax":
```
1 if True else pass
1 if True else return
1 if True else raise Exception('test')
1 if True else del smth
1 if True else yield 2
1 if True else assert False
1 if True else break
1 if True else continue
```

The following expressions will be handled by `a=invalid_ifexp_body_stmt 'if' b=disjunction 'else' c=expression` and generate the message "Syntax Error: statement given where 'body' expression required":
```
pass if True else 1
break if True else 1
continue if True else 1
```

The followingexpressions will be handled by `a=invalid_ifexp_body_stmt 'if' b=disjunction 'else' c=invalid_ifexp_orelse_stmt` and generate the message  "Syntax Error: statement given where 'body' expression required":
```
pass if True else pass
break if True else continue
```

sergey-miryanov committed:
gh-129515: Clarify syntax error messages for conditional expressions

sergey-miryanov commented on Grammar/python.gram:
```diff
@@ -114,6 +114,9 @@ simple_stmt[stmt_ty] (memo):
     | e=star_expressions { _PyAST_Expr(e, EXTRA) }
+    | simple_stmt_without_expressions
+
+simple_stmt_without_expressions[stmt_ty]:
```
I'm not sure if we should handle just 'pass' or other statements too. So I did it for others, but this is a discussion.

sergey-miryanov commented on Grammar/python.gram:
```diff
@@ -1184,6 +1187,9 @@ invalid_expression:
         RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "invalid syntax. Perhaps you forgot a comma?") }
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' c=simple_stmt_without_expressions { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (c, "statement given where 'orelse' expression required")}
+   | a=simple_stmt_without_expressions 'if' b=disjunction 'else' c=expression { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (a, "statement given where 'body' expression required")}
```
this is for 'pass if 1 else set()'

sergey-miryanov commented on Grammar/python.gram:
```diff
@@ -1184,6 +1187,9 @@ invalid_expression:
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' c=simple_stmt_without_expressions { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (c, "statement given where 'orelse' expression required")}
+   | a=simple_stmt_without_expressions 'if' b=disjunction 'else' c=expression { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (a, "statement given where 'body' expression required")}
+   | a=simple_stmt_without_expressions 'if' b=disjunction 'else' c=simple_stmt_without_expressions { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (a, "statement given where 'body' expression required")}
```
this is for 'continue if True else break'

sergey-miryanov committed:
Update news

sergey-miryanov committed:
Replace simple_stmt_without_expressions with explicit invalid_ifexp_*

sergey-miryanov commented on Grammar/python.gram:
```diff
@@ -1177,13 +1186,31 @@ invalid_type_param:
                 : "cannot use bound with ParamSpec")
         }
 
+invalid_ifexp_orelse_stmt[stmt_ty]:
```
I'm not sure if we should handle just 'pass' or other statements too. So I did it for others, but this is a discussion.

sergey-miryanov commented on Grammar/python.gram:
```diff
@@ -1177,13 +1186,31 @@ invalid_type_param:
         _PyPegen_check_legacy_stmt(p, a) ? NULL : p->tokens[p->mark-1]->level == 0 ? NULL :
         RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "invalid syntax. Perhaps you forgot a comma?") }
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' c=invalid_ifexp_orelse_stmt { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (c, "statement given where 'orelse' expression required")}
```
This is for '1 if 1 else pass'

sergey-miryanov commented on Grammar/python.gram:
```diff
@@ -1177,13 +1186,31 @@ invalid_type_param:
         RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "invalid syntax. Perhaps you forgot a comma?") }
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' c=invalid_ifexp_orelse_stmt { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (c, "statement given where 'orelse' expression required")}
+   | a=invalid_ifexp_body_stmt 'if' b=disjunction 'else' c=expression { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (a, "statement given where 'body' expression required")}
```
This is for 'pass if 1 else 1'

sergey-miryanov commented on Grammar/python.gram:
```diff
@@ -1177,13 +1186,31 @@ invalid_type_param:
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' c=invalid_ifexp_orelse_stmt { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (c, "statement given where 'orelse' expression required")}
+   | a=invalid_ifexp_body_stmt 'if' b=disjunction 'else' c=expression { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (a, "statement given where 'body' expression required")}
+   | a=invalid_ifexp_body_stmt 'if' b=disjunction 'else' c=invalid_ifexp_orelse_stmt { RAISE_SYNTAX_ERROR_KNOWN_LOCATION (a, "statement given where 'body' expression required")}
```
This is for 'continue if 1 else break'

sergey-miryanov committed:
Merge branch 'main' into gh-129515-ifexp-syntax-error

sergey-miryanov committed:
elif after else

picnixz commented on Lib/test/test_ast/test_ast.py:
```diff
@@ -1966,6 +1966,40 @@ def test_ifexp(self):
         for args in (s, l, l), (l, s, l), (l, l, s):
             self.expr(ast.IfExp(*args), "must have Load context")
 
+    def test_ifexp_orelse_stmt(self):
```
You could refactor the test by using something like:

```py
msg = "statement given where 'orelse' expression required"
for stmt in [
    "...",
    ...
]:
    self.subTest(stmt), self.assertRaisesRegex(SytanxError, msg):
        ast.parse(stmt)
```

Ditto for the other tests.

sergey-miryanov replied:
Fixed

sergey-miryanov committed:
Merge branch 'gh-129515-ifexp-syntax-error' of github.com:sergey-miryanov/cpython into gh-129515-ifexp-syntax-error

sergey-miryanov committed:
Merge branch 'main' into gh-129515-ifexp-syntax-error

sergey-miryanov committed:
Simplify test_ifexp_* tests

sergey-miryanov committed:
Remove an accidentally added test line

pablogsal committed:
Merge branch 'main' into gh-129515-ifexp-syntax-error

pablogsal commented on Grammar/python.gram:
```diff
@@ -1177,13 +1186,31 @@ invalid_type_param:
                 : "cannot use bound with ParamSpec")
         }
 
+invalid_ifexp_orelse_stmt[stmt_ty]:
```
This is going to add a heavy amount of backtracking to other error messages so we need to ensure that it only triggers with this subtree

sergey-miryanov replied:
Fixed, I think the current version is a little better.

picnixz commented on Lib/test/test_ast/test_ast.py:
```diff
@@ -1966,6 +1966,41 @@ def test_ifexp(self):
+            "x = 1 if 1 else pass",
+            "x = 1 if 1 else return",
+            "x = 1 if 1 else raise Exception('a')",
+            "a = 1; x = 1 if 1 else del a",
```
```suggestion
            "x = 1 if 1 else del a",
```

picnixz replied:
`ast.parse` doesn't need the variables to exist since it's only a parsing step. There is no evaluation of the code yet.

sergey-miryanov replied:
Fixed

picnixz commented on Misc/NEWS.d/next/Core_and_Builtins/2025-02-09-11-30-38.gh-issue-129515.3L3vmo.rst:
```diff
@@ -0,0 +1,2 @@
+Clarify syntax error messages for conditional expressions if statement
```
This is kind of a new feature so a What's New entry would also be great. It could be put in the same section where we improved `ValueError` messages, and it would also be good to actually show some examples (`body` and `orelse` parts are not necessarily known).

picnixz replied:
In addition, you could also link to the grammar specification and use the `:keyword:` role for `if`.

sergey-miryanov replied:
I tried to update the "What's New" section, but I'm not the best at writing in English.

pablogsal reviewed (changes requested):
I am not sure this is the right approach. I think a much cleaner approach is isolating the if expression in a rule and add an invalid rule that matches anything other than an expression at the end and we raise in that case.

The problem with the current approach is that in involves a heavy amount of backtracking for the invalid rules and also needs to manually maintain a list of 'all things that are invalid', which is not very clean, specially in this case.

Maybe I am missing something here though 🤔

sergey-miryanov committed:
Merge branch 'main' into gh-129515-ifexp-syntax-error

sergey-miryanov committed:
Simplify invalid_expression rule

sergey-miryanov committed:
Update whatsnew for 3.14

sergey-miryanov committed:
Fix references in whatsnew

sergey-miryanov committed:
Fix reference to if_expr

sergey-miryanov reviewed (commented):
Fixed in latest changes.

sergey-miryanov committed:
Merge branch 'main' into gh-129515-ifexp-syntax-error

sergey-miryanov committed:
Fix code-block in whatsnew

sergey-miryanov reviewed (commented):
I can't figure out how to mark that requested changes made.

pablogsal commented on Doc/whatsnew/3.14.rst:
```diff
@@ -174,6 +174,27 @@ Improved error messages
+  is passed to the :ref:`if_expr`, then the error message highlights where
+  the :token:`~python-grammar:expression` is required.
+
+  .. code-block:: pycon
```
```suggestion
  .. code-block:: python
```

sergey-miryanov replied:
Fixed, thanks!

pablogsal commented on Lib/test/test_ast/test_ast.py:
```diff
@@ -1966,6 +1966,42 @@ def test_ifexp(self):
+        ]:
+            with self.subTest(stmt), self.assertRaisesRegex(SyntaxError, msg):
+                ast.parse(stmt)
+
```
These test should not be here, you need to modify `test_syntax`

sergey-miryanov replied:
Fixed.

pablogsal commented on Grammar/python.gram:
```diff
@@ -1184,6 +1193,11 @@ invalid_expression:
         RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "invalid syntax. Perhaps you forgot a comma?") }
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' c[stmt_ty]=(
+        return_stmt | raise_stmt | pass_stmt | del_stmt | yield_stmt | assert_stmt | break_stmt | continue_stmt) {
```
This has the same problem as I mentioned before: we need to list all the different statements, which I don't think we should be doing.

sergey-miryanov replied:
Thanks, fixed as you suggested.

pablogsal commented on Grammar/python.gram:
```diff
@@ -1184,6 +1193,11 @@ invalid_expression:
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' c[stmt_ty]=(
+        return_stmt | raise_stmt | pass_stmt | del_stmt | yield_stmt | assert_stmt | break_stmt | continue_stmt) {
+        RAISE_SYNTAX_ERROR_KNOWN_LOCATION (c, "statement given where 'orelse' expression required") }
```
Something like this:


```suggestion
      | a=disjunction 'if' b=disjunction 'else' !expression {
        RAISE_SYNTAX_ERROR_ON_NEXT_TOKEN("statement given where 'orelse' expression required") }
```

pablogsal replied:
But the error is also not very clear: 'orelse' means nothing to most users so we need to select something that highlights better what's missing

sergey-miryanov replied:
Oh, whoa! Sorry, I didn't see the forest for the trees.

sergey-miryanov replied:
Thanks for pointing me a right solution! And thanks for patience! I didn't know about RAISE_SYNTAX_ERROR_ON_NEXT_TOKEN, but now I know a bit more.

Also, I tried to reword error messages (I didn't find out how those parts of if-else expressions were named but searched through the codebase).

pablogsal reviewed (changes requested):
Left some new comments

sergey-miryanov committed:
Merge branch 'main' into gh-129515-ifexp-syntax-error

sergey-miryanov committed:
Fix invalid rule for ifexp if statement given after else

- also reword error messages
- move tests to test_syntax
- update whatsnew and news

sergey-miryanov committed:
Update doc-string in test_syntax

StanFromIreland commented on Doc/whatsnew/3.14.rst:
```diff
@@ -175,6 +175,31 @@ Improved error messages
+  :keyword:`yield`, :keyword:`raise`, :keyword:`break`, :keyword:`continue`,
+  :keyword:`assert`, :keyword:`import`, :keyword:`from`) is passed to the
+  :ref:`if_expr` after :keyword:`else`, or one of the :keyword:`pass`,
+  :keyword:`break`, :keyword:`continue` is passed before :keyword:`if`, then the
```
```suggestion
  :ref:`if_expr` after :keyword:`else`, or one of :keyword:`pass`,
  :keyword:`break`, or :keyword:`continue` is passed before :keyword:`if`, then the
```

sergey-miryanov replied:
Thanks!

sergey-miryanov committed:
Update Doc/whatsnew/3.14.rst

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

pablogsal reviewed (approved):
Ah, this looks MUCH better. Really well done @sergey-miryanov! :+1:

pablogsal commented on Grammar/python.gram:
```diff
@@ -1187,6 +1196,10 @@ invalid_expression:
    | a=disjunction 'if' b=disjunction !('else'|':') { RAISE_SYNTAX_ERROR_KNOWN_RANGE(a, b, "expected 'else' after 'if' expression") }
+   | a=disjunction 'if' b=disjunction 'else' !expression {
+        RAISE_SYNTAX_ERROR_ON_NEXT_TOKEN("expected expression after 'else', but statement is given") }
+   | a[stmt_ty]=(pass_stmt|break_stmt|continue_stmt) 'if' b=disjunction 'else' c=simple_stmt {
```
The only thing I don't really like is the `(pass_stmt|break_stmt|continue_stmt)`. Could we maybe use just `!expression` or `simple_stmt` if you want to capture it?

sergey-miryanov replied:
If I use `a=simple_stmt 'if' b=disjunction 'else' c=simple_stmt` then I get messages like `SyntaxError: cannot delete conditional expression` for `x = del a if True else 1`.
If I use `!expression 'if' b=disjunction 'else' c=simple_stmt` then it breaks unrelated tests, and I'm not so confident to touch them.
If I explicitly add del_stmt to alternatives like `a[stmt_ty]=(pass_stmt|break_stmt|continue_stmt|del_stmt) 'if' b=disjunction 'else' c=simple_stmt` I also get `SyntaxError: cannot delete conditional expression`.
So, I decided to enumerate only those statements that I think users can try to write and that looks legit.

pablogsal replied:
Works for me as a compromise. Thanks for trying it out!

picnixz commented on Doc/whatsnew/3.14.rst:
```diff
@@ -175,6 +175,31 @@ Improved error messages
+  error message highlights where the :token:`~python-grammar:expression` is
+  required. (Contributed by Sergey Miryanov in :gh:`129515`.)
+
+  .. code-block:: python
```
```suggestion
  .. code-block:: pycon
```

sergey-miryanov replied:
@pablogsal suggested to use python here, so I'm not sure what actual I should use? And I can't find what to use in devguide (maybe I'm oversight something).

picnixz replied:
Sphinx is smart enough to detect that `>>>` is for python console *assuming* that the current syntax highlighting is `python` (which is the default for the file I think, so `python` is not even needed). However, I think it's better to use an explicit `pycon` lexer when possible.

sergey-miryanov replied:
Thanks for the explanation.

picnixz commented on Lib/test/test_syntax.py:
```diff
@@ -2863,6 +2875,34 @@ def test_match_stmt_invalid_as_expr(self):
         )
 
+    def test_ifexp_else_stmt(self):
+        msg = "expected expression after 'else', but statement is given"
```
This is really some nitpick but you could have:

```py
msg = ...
check = lambda s: self._check_error(s, msg)

check(...)
check(...)
```

That way you avoid rewriting `, msg` everytime. Or you can use the loop approach I suggested:

```py
msg = ...
for rhs in [

]:
    self._check_error(f"x = 1 if 1 else {rhs}", msg)
```

sergey-miryanov replied:
I tried to mimic style from the other tests in the file. But I think the code will be much clearer if I use your suggestions. Thanks!

sergey-miryanov replied:
Fixed

picnixz commented on Misc/NEWS.d/next/Core_and_Builtins/2025-02-09-11-30-38.gh-issue-129515.3L3vmo.rst:
```diff
@@ -0,0 +1,2 @@
+Clarify syntax error messages for conditional expressions if statement
+specified before ``if`` or after ``else`` keywords.
```
```suggestion
Clarify syntax error messages for conditional expressions when a statement
is specified before an :keyword:`if` or after an :keyword:`else` keyword.
```

sergey-miryanov replied:
Thanks!

sergey-miryanov committed:
Update Misc/NEWS.d/next/Core_and_Builtins/2025-02-09-11-30-38.gh-issue-129515.3L3vmo.rst

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

sergey-miryanov committed:
Update tests and whatsnew entry

sergey-miryanov committed:
Merge branch 'gh-129515-ifexp-syntax-error' of github.com:sergey-miryanov/cpython into gh-129515-ifexp-syntax-error
